### ISSUE TITLE
Inconsistent/unnecessary escaping of characters in inputs

### ISSUE THREAD

**Issue by nikku** [camunda/zeebe#9859]
**Describe the bug**

As a user I want to pass special chars in strings ([FEEL ref](https://docs.camunda.io/docs/components/modeler/feel/language-guide/feel-data-types/#string)) in a way that I can consume them in a job worker.

I've defined the following input mappings:

| name | value |
|:---|:---|
| dynamic_input | `= "Hello\n\nYOU!"` |
| static_input | `Hello\n\nYOU!` |
| static_input_explicit_newlines | `HELLO<newline><newline>YOU!` |

Executing my process I receive the following input variable values in operate (but also in my process worker):

| name | value |
|:---|:---|
| dynamic_input | `"Hello\\n\\nYOU!"` |
| static_input | `"Hello\\\\n\\\\nYOU!"` |
| static_input_explicit_newlines | `"Hello\\n\\nYOU!"` |

As you can see inputs are escaped differently depending on whether I use `\n` encoded newlines or actual newline characters. Escaping happening in the first place is fine I believe as this is a JSON string and `\\n` is the standard way to encode newlines there.

**To Reproduce**

* Deploy + run [test process](https://github.com/camunda/zeebe/files/9159826/input-test.bpmn.txt)
* See double escaping happening in string inputs

**Expected behavior**

We do the unescaping magic ourselves so that in user land no one needs to worry:

  * I define an input mapping `foo = "a\nb"`
  * In my job worker I see `foo = "a\nb"`

As a more user un-friendly alternative we could consider: String inputs are escaped like expression inputs (no double escaping), we clearly document how escaped strings shall be worked with (and how un-escaping has to be done in user land). This is closer to the existing behavior.

**Log/Stacktrace**

none.

**Environment:**
- OS: Linux
- Zeebe Version: 8.0.2
- Configuration: nothing special

Support: 
[SUPPORT-15887](https://jira.camunda.com/browse/SUPPORT-15887)
[SUPPORT-17621](https://jira.camunda.com/browse/SUPPORT-17621)

**Comment by nikku** [comment]
To add to this I see zeebe escaping unicode characters (i.e. `🤩`) despite JSON having full unicode support.

```
Hello\n\nYOU! 🤩
```

turns into 

```
"Hello\\n\\nYOU! \uD83E\uDD29"
```

**Comment by nikku** [comment]
Is there a chance this can be fixed at some point? Or do we regard this as a nasty-bug-turned-public-api?

**Comment by menski** [comment]
@felix-mueller this is something which mostly effects connectors but it is very unexpected behavior. We would like to understand if you have any input on priority for this.

**Comment by felix-mueller** [comment]
It should be fixed at some point. Since we still need to root-cause and better understand the issue it is probably not S. We won't priotize it for the next few iterations (first we need to finish instance modification, DMN evaluation and delete definitions).

**Comment by menski** [comment]
Thanks Felix, I put the priority on "Later" for now

**Comment by darox** [comment]
I don't know whether I'm hitting the same issue. I'm using the Go client to ingest JSON formatted data into Zeebe. Jobs are failing due to escaped double quotes. If I set vars manually via simple-monitor everything works as expected.

**Comment by igpetrov** [comment]
Hi team, is there any ETA to fix this? It is affecting currently 3 connectors.

**Comment by felix-mueller** [comment]
@igpetrov currently we have planned to look into it after instance modification, DMN evaluation and Delete Definitions. these topics are scheduled to be included in our April minor release, so if we don't change priorities we probably will look at it in a few months.

from connectors perspective can we live with this for a few months?

**Comment by igpetrov** [comment]
Hi @felix-mueller, appreciate your update. Absolutely. Just wanted to understand the ETA. Thank you!

**Comment by tmetzke** [comment]
For reference, in Connectors, we currently do the following to work around this with data we receive as variables after Zeebe has processed the user input:
```
StringEscapeUtils.unescapeJson(body.toString())
```
This uses `StringEscapeUtils` from `org.apache.commons:commons-text`.

**Comment by nikku** [comment]
@felix-mueller The longer we keep such behavior in the core, the harder it will be for people to unbuild their hacks (trying to work with zeebe data). My :two: :coin:.

**Comment by nikku** [comment]
A similar issue is reported by a customer through [SUPPORT-15887](https://jira.camunda.com/browse/SUPPORT-15887).

**Comment by menski** [comment]
Discussed in the planning meeting on 2023-02-17. Right now we don't have capacity to work on this, and deprioritize it as the customer has a workaround available.

**Comment by korthout** [comment]
When we work on this, we need to make sure not to break user space when backporting it to patches. We can do this by adding a configuration setting that allows switching between the current (broken) and new (fixed) behaviors. The default should stay the same in patch releases, but in a minor release we can change the default. This should be clearly communicated in the release notes (and blog) so users understand the change.

**Comment by abbasadel** [comment]
This issue was mentioned again in https://jira.camunda.com/browse/SUPPORT-17621

**Comment by korthout** [comment]
When we resolve this issue, please verify whether the fix also resolves:
- https://github.com/camunda/zeebe/issues/13551

**Comment by Benjoyo** [comment]
Also came across this, the `StringEscapeUtils.unescapeJson` workaround works for now but would be nice to have this fixed soon :)

**Comment by mschoe** [comment]
I run into this issue when using the GH outbound connector and create a new issue with a multi line description field.  As this is a quite common use case I'm wondering when we are planning to provide a fix. @felix-mueller is this still scheduled for the 8.5 release in April 2024?

**Comment by christian-konrad** [comment]
Would https://github.com/camunda/product-hub/issues/1040 help here or is it just a quick fix on string behavior?

**Comment by saig0** [comment]
@nicpuppa started to work on this issue by fixing the escaping in the FEEL engine [here](https://github.com/camunda/feel-scala/pull/750). We plan to deliver a fix in one of the next patch releases. :rocket: 

@christian-konrad the FEEL templating is a bigger topic. Let's fix the issue with the escaping first. :+1:

**Comment by christian-konrad** [comment]
Perfect @saig0 ! Thanks

**Comment by MaxTru** [comment]
Hi @abbasadel ,

according to the [previous assessment](https://github.com/camunda/zeebe/issues/9859#issuecomment-1805096523) this should have been fixed with this [version bump](https://github.com/camunda/zeebe/commit/4a90c77137dcad51106ee1e2fa8d1172c12640c2). Can you verify from Zeebe perspective please and update this issue? Thanks

**Comment by megglos** [comment]
We need to clarify whether this change is breaking, especially in regards to present workaround in other components such as https://github.com/camunda/connector-slack/issues/48 .

If it's breaking we may prefer to only release it in the next minor and need to coordinate with the affected components to remove the workaround applied.

@abbasadel will sync with @saig0 on this tomorrow and ping back.

**Comment by abbasadel** [comment]
Hi everyone, this is Zeebe's assessment: 
- The recent [bug fix](https://github.com/camunda/feel-scala/pull/750) on the FEEL engine is not backward compatible from Zeebe's point of view. However, we believe, based on @saig0’s feedback and our tests, the impact on our users is low since the recommended workaround we provided was to use [unescapeJson API](https://commons.apache.org/proper/commons-lang/apidocs/org/apache/commons/lang3/StringEscapeUtils.html#unescapeJson-java.lang.String-) from Apache Commons which handles breaking change.
- As the previous behavior was wrong, mentioning the new behavior change in the release notes and as an announcement would be needed
- This FEEL fix should resolve [bug](https://github.com/camunda/connectors/issues/239), to be validated by Connectors/Operate teams. 
- The support team should also be aware of this FEEL behavior change once release

**Comment by MaxTru** [comment]
> Hi everyone, this is Zeebe's assessment:
> 
> * The recent [bug fix](https://github.com/camunda/feel-scala/pull/750) on the FEEL engine is not backward compatible from Zeebe's point of view. However, we believe, based on @saig0’s feedback and our tests, the impact on our users is low since the recommended workaround we provided was to use [unescapeJson API](https://commons.apache.org/proper/commons-lang/apidocs/org/apache/commons/lang3/StringEscapeUtils.html#unescapeJson-java.lang.String-) from Apache Commons which handles breaking change.
> * As the previous behavior was wrong, mentioning the new behavior change in the release notes and as an announcement would be needed
> * This FEEL fix should resolve [bug](https://github.com/camunda/connectors/issues/239), to be validated by Connectors/Operate teams.
> * The support team should also be aware of this FEEL behavior change once release

Thanks @abbasadel . Will you and the ZPA team please ensure that all these things happen? If so, can we please track this via a breakdown and DRI assignments? Why is the issue not in `in progress`?

**Comment by abbasadel** [comment]
@MaxTru, we will take care of this issue

**Comment by nicpuppa** [comment]
Hi everyone, as follow up I created 3 separate issues to reduce the scope and make easier to follow the progress:

- [Verify correct behaviour of input mappings after bump to FEEL 1.17.3](https://github.com/camunda/zeebe/issues/15445) -> this is a testing purpose issue, to verify that the new FEEL version has the correct behaviour
- [Special characters in static inputs should not be escaped](https://github.com/camunda/zeebe/issues/15447) -> as mentioned also by @saig0 in [this comment](https://github.com/camunda/feel-scala/pull/750#pullrequestreview-1731283017), characters in static input are still escaped. Static input are handled directly by Zeebe
- [Verify unicode characters are not escaped in input mappings](https://github.com/camunda/zeebe/issues/15446) -> after resolving the [feel issue related to unicode](https://github.com/camunda/feel-scala/issues/558), on Zeebe we should verify that the behaviour is as expected (unicode characters should not be escaped)

In accordance with @abbasadel and @saig0 I will close this issue

**Comment by abbasadel** [comment]
Thanks, @nicpuppa, for the updates. 

I added the sub-issues into the main issue description for clarity and tracking purpose

**Comment by abbasadel** [comment]
Hi everyone, 
- The [dynamic_input unnecessary escaping bug](https://github.com/camunda/feel-scala/pull/750) is currently resolved in [Zeebe 8.3.4 release](https://github.com/camunda/zeebe/releases/tag/8.3.4)
- The bug fix is https://github.com/camunda/zeebe/issues/15445 to resolve the [connectors team bug](https://github.com/camunda/connectors/issues/239). Thanks @saig0 and @nicpuppa 
- The corrected behavior change is highlighted in the [release notes](https://github.com/camunda/zeebe/releases/tag/8.3.4). Thanks @saig0 for the wording.
- The support team is informed about the [corrected behavior change](https://camunda.slack.com/archives/CHAC0L80M/p1702029188512579)
- Next action: Updated [input mappings docs](https://docs.camunda.io/docs/components/concepts/variables/#input-mappings) as suggested by @akeller

**Comment by akeller** [comment]
@abbasadel will this be included with the alpha release next week? Since this is a Zeebe task, I'm not sure what the expectation is. When you open your docs PR, please follow our PR template so we can review ASAP if you want it reviewed, merged, and published on Tuesday.

**Comment by abbasadel** [comment]
Thanks @akeller. The docs [PR](https://github.com/camunda/camunda-docs/pull/3020) is already merged

**Comment by nicpuppa** [comment]
> It seems this can be closed as done? @nicpuppa ?

Sorry, I missed the comment @npepinpe 🙇 . There is still a [task](https://github.com/camunda/camunda/issues/15446) to be completed, that is blocked by [this](https://github.com/camunda/feel-scala/issues/558) FEEL issue
